File: 2025/day01.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <iterator>
#include <cstdint>
#include <cstring>
#include <cmath>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <numeric>

#include "utils.h"

using namespace std;

const string test(
    "L68\n"
    "L30\n"
    "R48\n"
    "L5\n"
    "R60\n"
    "L55\n"
    "L1\n"
    "L99\n"
    "R14\n"
    "L82"
);

#define DAY "day01"

bool DEBUG = false;
bool TEST = false;

int part1 (StringVector & data )
{
    int count = 0;
    int pos = 50;
    for( auto & line : data )
    {
        int c = std::stoi( line.substr( 1 ) );
        if( line[0] == 'L' )
            pos = (pos - c + 100) % 100;
        else
            pos = (pos + c) % 100;
        count += !pos;
    }
    return count;
}

int part2 (StringVector & data)
{
    int count = 0;
    int pos = 50;
    for( auto & line : data )
    {
        int c = std::stoi( line.substr( 1 ) );
        if( line[0] == 'L' )
        {
            count -= !pos;
            pos -= c;
            while( pos < 0 )
            {
                count++;
                pos += 100;
            }
            count += !pos;
        }            
        else
        {
            pos += c;
            count += pos / 100;
            pos %= 100;
        }
    }
    return count;
}


int main( int argc, char ** argv )
{
    string name = *argv;
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    string input = TEST ? test : file_contents(DAY".txt");
    StringVector data = split(input);
    
    cout << "Part 1: " << part1(data) << "\n";
    cout << "Part 2: " << part2(data) << "\n";
}

File: 2025/day01.go
package main

import (
	"fmt"
	"strings"

	"aoc/tools"
	_ "embed"
)

var DEBUG bool = false
var TEST bool = false
var test = `
L68
L30
R48
L5
R60
L55
L1
L99
R14
L82`[1:]

//go:embed day01.txt
var live string

func part1(data []string) int {
	pos := 50
	count := 0
	for _, line := range data {
		v := tools.StrToInt(line[1:len(line)])
		if line[0] == 'L' {
			pos = (pos - v + 100) % 100
		} else {
			pos = (pos + v) % 100
		}
		if pos == 0 {
			count++
		}
		if DEBUG {
			fmt.Println(line, pos, count)
		}
	}
	return count
}

func part2(data []string) int {
	pos := 50
	count := 0
	for _, line := range data {
		v := tools.StrToInt(line[1:len(line)])
		if line[0] == 'L' {
			if pos == 0 {
				count--
			}
			pos -= v
			for pos < 0 {
				pos += 100
				count++
			}
			if pos == 0 {
				count++
			}
		} else {
			pos += v
			count += pos / 100
			pos %= 100
		}
		if DEBUG {
			fmt.Println(line, pos, count)
		}
	}
	return count
}

func main() {
	var input string
	TEST, DEBUG, input = tools.Setup(test, live)
	data := strings.Split(input, "\n")

	fmt.Println("Part 1:", part1(data))
	fmt.Println("Part 2:", part2(data))
}
